src/CakeSlayer/Measure.hs

```diff
@@ -0,0 +1,91 @@
+{-# LANGUAGE BangPatterns #-}
+
+{- | This module contains the 'MonadMeasure' type class which purpose is to
+apply some metric calculations on actions, e.g. taken time.
+-}
+
+module CakeSlayer.Measure
+       ( MonadMeasure
+       , timedAction
+
+         -- * Internals
+       , timedActionImpl
+       ) where
+
+import Prometheus (Histogram, Info (..), MonadMonitor, defaultBuckets, histogram, observe, register)
+import Relude.Extra.CallStack (ownName)
+import System.CPUTime (getCPUTime)
+import System.Metrics.Distribution (Distribution)
+
+import CakeSlayer.Has (Has, grab)
+
+import qualified Data.HashMap.Strict as HashMap
+import qualified Data.Map as Map
+import qualified System.Metrics as Metrics
+import qualified System.Metrics.Distribution as Distribution
+
+{- | This type class is used to perform action and measure time and other
+metrics needed to be performed.
+-}
+class Monad m => MonadTimed m where
+    timedAction :: HasCallStack => m a -> m a
+
+
+-- QUESTION: move to CakeSlayer.Env ?
+type Timings = IORef (Map Text Distribution)
+type PrometheusRegistry = IORef (HashMap Text Histogram)
+
+type MonadMeasure m = (HasCallStack, MonadTimed m)
+
+{- | Helper function to be used in instance implementations.
+
+It measures the time taken to perform the given action and store it
+in the @timings@ distribution with the label of the action.
+
+This is a recommended implementation which has the integration with @prometheus@.
+-}
+timedActionImpl
+    :: forall r m a .
+       ( MonadReader r m
+       , Has Timings r
+       , Has Metrics.Store r
+       , Has PrometheusRegistry r
```

Comment:
I'm not sure, what is the best here: do it with such constraints (like it is right now) or just use it as function arguments, so the one who would use it in the app should do all the `grab` etc. stuff in the instance implementation?

Reply:
@vrom911 I think that doing it with constraints is fine and in line with our architecture. But we need to think, how to refactor and separate this function into two, so users can choose whatever they want...